tracing/src/main/java/com/palantir/tracing/TraceLocal.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public final class TraceLocal<T> {
+
+    private static final SafeLogger log = SafeLoggerFactory.get(TraceLocal.class);
+
+    private final Supplier<T> initialValue;
+
+    /**
+     * Creates a trace local variable.
+     *
+     * When not already set, and this variable is accessed in a trace with the {@link #get()} method, the supplier
+     * will be invoked to supply a value (which will then be stored for future accesses).
+     *
+     * The supplier is thus normally invoked once per trace, but may be invoked again in case of subsequent
+     * invocations of {@link #remove()} followed by get.
+     */
+    public TraceLocal(Supplier<T> initialValue) {
+        this.initialValue = initialValue;
+    }
+
+    /**
+     * Retrieve the current value of the trace local, with respect to the current trace.
+     *
+     * If there is no current trace, i.e. {@link Tracer#hasTraceId()} is null, then this will return null.
+     *
+     * If the value of this trace local has not been set for the current trace, then the supplier passed in the
+     * constructor will be called to supply a value.
+     */
+    @Nullable
+    public T get() {
+        return Tracer.getTraceLocalValue(this, initialValue);
+    }
+
+    /**
+     * Sets the value of this trace local for the current trace.
+     */
+    public void set(@Nonnull T value) {
```

Comment:
Unlike ThreadLocal, this can race because traces are shared across threads. I wonder if we should provide some way to acknowledge replaced values by returning the previous value or taking some form of collector. Likely unnecessary in most cases, but it would be helpful to retain the capability.

Reply:
what do you mean by taking a collector?

Reply:
either we return the previous value from `set`, or allow you to provide an update function which may take the previous value e.g. `(TraceLocal<Integer>) traceLocal.update(1, (previousValue, newValue) -> previous == null ? newValue : previousValue + newValue);` I think this sort of API is overkill though.

Edge cases returning the original value from `set` include:
* Do we return `initialValue.get()` if the value is unset?
* Should we provide an API to safely update a value if the current value matches our expectation (avoid update races)?
* Do we need a `setIfAbsent` for the common case, to avoid clobbering values across threads? Otherwise setting, and unsetting when our expectations arent met could allow mutations to be consumed by the temporary update value that we roll back.

Reply:
will return original value from set for now (documenting that this returns null if unset, not initialValue.get()). We can add update/setIfAbsent stuff in future if needed.

Reply:
(as in, those are easy additive changes)